GRC Updates - Draft 2

dhaiducek opened the pull request:
Alternative to:
- #3857

Addresses:
- https://github.com/stolostron/backlog/issues/23525
- [Bug 2090781](https://bugzilla.redhat.com/show_bug.cgi?id=2090781) - [RFE DOC] Highlight difference between "musthave" and "mustonlyhave"

Closes #3857

dhaiducek committed:
Polish docs

- Use fully qualified API endpoint
- Polish descriptions
- Remove custom controller refs
- Add LabelSelector for NamespaceSelector

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek committed:
Update ConfigurationPolicy docs

- Remove repetitive policy lists
- Add policy table
- Add additional links

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek committed:
Add generator links

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek committed:
Make sample policy pages consistent

- Link to parent policy docs
- Use the same description table
- Strip out optional values in the YAML

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dockerymick committed:
Polish docs

- Minor polish updates
- Monospace fields in tables
- Updated link references
- update after Dale provided feedback
- Moved the xref and external links to the end of the page

dockerymick commented on governance/cert_policy_ctrl.adoc:
```diff
@@ -67,29 +68,16 @@ spec:
-
-If the `namespaceSelector` for the certificate policy controller does not match any namespace, the policy is considered compliant.
+| Determines namespaces in the managed cluster where secrets are monitored. The `include` and `exclude` parameters accept file path expressions to include and exclude namespaces by name. The `matchExpressions` and `matchLabels` parameters specify namespaces to be included by label. See the link:https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/[Kubernetes labels and selecto ...
+//^instead on ANDing how about "combining". Is ANDing the name of a task? Will the user ever see "AND" as an option?
```
@dhaiducek

dhaiducek replied:
I suppose "ANDing" isn't a word, huh? 😉 

How about "The resulting list is compiled by using the intersection of results from all parameters."?

For background, "combining" would work, but I'd like to make it clear that if a namespace is selected in by one selector but not by another, it won't be included.

dockerymick commented on governance/image_vuln_policy.adoc:
```diff
@@ -10,147 +10,90 @@ For more information about the Security Operator, see the _Container Security Op
 
 * <<image-vulnerability-policy-yaml-structure,Image vulnerability policy YAML structure>>
-* <<image-vulnerability-policy-yaml-table,Image vulnerability policy YAML table>>
+//* <<image-vulnerability-policy-yaml-table,Image vulnerability policy YAML table>>
```
@dhaiducek Can you clarify why the table was removed. Just want to be on the same page.

dhaiducek replied:
This policy is very similar to the Compliance Operator, so I was thinking the policies that installed Operators could follow the same patterns.

dockerymick replied:
@dhaiducek Thanks for clarifying and pointing that out. I didn't realize that there wasn't a table. Should we add a parameter table to the compliance operator policy doc?

dhaiducek replied:
We have some comments in the image vulnerability policy here that we could add there also that I think make things clear.

I'm of the opinion that the ConfigurationPolicy YAML is documented elsewhere and the Compliance Operator is also documented elsewhere and we provide links to them, so I'd personally want users to use those instead of pursuing our own table.

dockerymick replied:
Thanks! What should be added?

dhaiducek commented:
@dockerymick The rebase to resolve conflicts got really messy. I think I captured all of the changes that were made, but we may both need to take another look to make sure I resolved the merge conflicts correctly. Sorry!

dhaiducek committed:
Monospace fields in tables

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek commented on governance/etcd_encryption_policy.adoc:
```diff
@@ -19,85 +19,75 @@ Your `etcd-encryption` policy might resemble the following YAML file:
-See https://github.com/stolostron/policy-collection/blob/main/stable/SC-System-and-Communications-Protection/policy-etcdencryption.yaml[`policy-etcdencryption.yaml`] for the policy sample. See xref:../governance/create_policy.adoc#managing-security-policies[Managing security policies] for more information.
-
-Refer to xref:../governance/config_policy_ctrl.adoc#kubernetes-configuration-policy-controller[Kubernetes configuration policy controller] to view other configuration policies that are monitored by the controller.
+See https://github.com/stolostron/policy-collection/blob/main/stable/SC-System-and-Communications-Protection/policy-etcdencryption.yaml[`policy-etcdencryption.yaml`] for the policy sample. See xref:../governance/create_policy.adoc#managing-security-policies[Managing security policies] for more information. Refer to the xref:../governance/policy_example.adoc[Policy YAML documentation], and to xref ...
```
I added the links to highlight that there were more details and fields (like `namespaceSelector` and `evaluationInterval`). How about this wording?
```suggestion
See https://github.com/stolostron/policy-collection/blob/main/stable/SC-System-and-Communications-Protection/policy-etcdencryption.yaml[`policy-etcdencryption.yaml`] for the policy sample. See the xref:../governance/policy_example.adoc[Policy YAML documentation] and the xref:../governance/config_policy_ctrl.adoc#kubernetes-configuration-policy-controller[Kubernetes configuration policy controller] to view additional details on policy and configuration policy fields.
```

dockerymick committed:
Update governance/etcd_encryption_policy.adoc

Accepting suggestion

Co-authored-by: Dale Haiducek <dhaiduce@redhat.com>

dockerymick committed:
Update config_policy_ctrl.adoc

small update

dockerymick committed:
Update config_policy_ctrl.adoc

small changes

dockerymick committed:
Update config_policy_ctrl.adoc

dockerymick commented on governance/create_config_pol.adoc:
```diff
@@ -29,6 +18,9 @@ Learn to create, apply, view, and update your configuration policies. As a remin
 
 You can create a YAML file for your configuration policy from the command line interface (CLI) or from the console.
+
+If you have an existing Kubernetes manifest, the xref:../governance/policy_generator.adoc[policy generator] is also a useful tool to automatically wrap those manifests in a policy.
```
@dhaiducek I suggest that we write the following statement:

If you have an existing Kubernetes manifest, consider using the policy generator to automatically include the manifests in a policy. See the xref:../governance/policy_generator.adoc[Policy generator] documentation.

What do you mean when you say wrap? Is my suggestion somewhat accurate?

dhaiducek replied:
Yes, your suggestion looks good to me!

dockerymick committed:
Update etcd_encryption_policy.adoc

updates

dockerymick committed:
Update pod_policy.adoc

Moving external references to the end of the file

dockerymick committed:
Update create_config_pol.adoc

Updates after confirmation

dhaiducek commented:
/hold for edits and reviews

dockerymick committed:
Update image_vuln_policy.adoc

updates

dockerymick commented on governance/image_vuln_policy.adoc:
```diff
@@ -10,151 +10,91 @@ For more information about the Security Operator, see the _Container Security Op
----
+  remediationAction: enforce  # will be overridden by remediationAction in parent policy
+  severity: high
+  object-templates:
```
Let me know what else you might want to add in this doc

dockerymick committed:
Update etcd_encryption_policy.adoc

Naming the table

dockerymick committed:
Update memory_policy.adoc

updates

dockerymick committed:
Update policy_controllers.adoc

updates

dockerymick committed:
Update policy_example.adoc

Updates

dockerymick committed:
Update policy_generator.adoc

updates

dockerymick committed:
Update policy_sample_intro.adoc

Fixing link and other updates

dockerymick committed:
Update policy_sample_intro.adoc

couple more updates

dockerymick committed:
Update policy_set_ctrl.adoc

Updates

mprahl commented on governance/config_policy_ctrl.adoc:
```diff
@@ -1,65 +1,21 @@
 
-The configuration policy controller is created on the hub cluster during installation.
-Configuration policy controller supports the `enforce` feature and monitors the compliance of the following policies:
+The configuration policy controller is created on the managed cluster during installation. Configuration policy controller supports the `enforce` feature to remediate when the configuration policy is non-compliant. When the `remediationAction` for the configuration policy is set to `enforce`, the controller applies the specified configuration to the target managed cluster.
```
```suggestion
The configuration policy controller is created on the managed cluster during installation. The configuration policy controller supports the `enforce` feature to remediate when the configuration policy is non-compliant. When the `remediationAction` for the configuration policy is set to `enforce`, the controller applies the specified configuration to the target managed cluster.
```

mprahl commented on governance/config_policy_ctrl.adoc:
```diff
@@ -1,65 +1,21 @@
 
-The configuration policy controller communicates with the local Kubernetes API server to get the list of your configurations that are in your cluster.
-For more information about CRDs, see https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/[Extend the Kubernetes API with CustomResourceDefinitions].
+A Kubernetes object is defined (in whole or in part) in the `object-templates` array in the configuration policy, indicating to the configuration policy controller on the fields to compare with objects on the managed cluster. The configuration policy controller communicates with the local Kubernetes API server to get the list of your configurations that are in your cluster. For more information a ...
```
```suggestion
A Kubernetes object is defined (in whole or in part) in the `object-templates` array in the configuration policy, indicating to the configuration policy controller of the fields to compare with objects on the managed cluster. The configuration policy controller communicates with the local Kubernetes API server to get the list of your configurations that are in your cluster. For more information about CRDs, see https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/[Extend the Kubernetes API with CustomResourceDefinitions].
```

mprahl commented on governance/config_policy_ctrl.adoc:
```diff
@@ -73,24 +29,26 @@ spec:
-           - containerPort: 80
+  severity: low
   evaluationInterval:
     compliant:
```
Should these have values?

dockerymick replied:
No, we choose to have an empty policy example so that users could see the framework

mprahl commented on governance/config_policy_ctrl.adoc:
```diff
@@ -100,54 +58,55 @@ spec:
-| spec.remediationAction.complianceType
+| `spec.object-templates[].complianceType`
 | Required
 | Used to list expected behavior for roles and other Kubernetes object that must be evaluated or applied to the managed clusters. You must use the following verbs as parameter values:
```
```suggestion
| Used to define the desired state of the Kubernetes object on the managed clusters. You must use one of the following verbs as the parameter value:
```

mprahl commented on governance/config_policy_ctrl.adoc:
```diff
@@ -100,54 +58,55 @@ spec:
-| Specifies the severity when the policy is non-compliant. Use the following parameter values: `low`, `medium`, or `high`.
+| The array of Kubernetes objects (either fully defined or containing a subset of fields) for the controller to compare with objects on the managed cluster.
 
-| spec.remediationAction.complianceType
```
It'd be good to also document `metadataComplianceType`:
https://github.com/stolostron/config-policy-controller/blob/09e981ce5a35e6fb500b0c36962383b4cb18fa90/deploy/crds/policy.open-cluster-management.io_configurationpolicies.yaml#L140

dockerymick committed:
Update governance/config_policy_ctrl.adoc

Co-authored-by: Matt Prahl <mprahl@users.noreply.github.com>

dockerymick committed:
Update governance/config_policy_ctrl.adoc

Co-authored-by: Matt Prahl <mprahl@users.noreply.github.com>

dockerymick committed:
Update governance/config_policy_ctrl.adoc

Co-authored-by: Matt Prahl <mprahl@users.noreply.github.com>

mprahl commented on governance/config_policy_ctrl.adoc:
```diff
@@ -1,65 +1,21 @@
 = Kubernetes configuration policy controller
 
-Configuration policy controller can be used to configure any Kubernetes resource and apply security policies across your clusters.
+Configuration policy controller can be used to configure any Kubernetes resource and apply security policies across your clusters. The configuration policy is provided in the `policy-templates` field of the policy on the hub cluster, and is propagated to the selected managed clusters by the governance framework. See the xref:../governance/policy_example.adoc#policy-yaml-structure[policy YAML docu ...
```
```suggestion
The configuration policy controller can be used to configure any Kubernetes resource and apply security policies across your clusters. The configuration policy is provided in the `policy-templates` field of the policy on the hub cluster, and is propagated to the selected managed clusters by the governance framework. See the xref:../governance/policy_example.adoc#policy-yaml-structure[policy YAML documentation] for more about the hub cluster policy.
```

dockerymick committed:
Update governance/config_policy_ctrl.adoc

Co-authored-by: Matt Prahl <mprahl@users.noreply.github.com>

mprahl commented on governance/config_policy_ctrl.adoc:
```diff
@@ -1,65 +1,21 @@
 
-The configuration policy controller communicates with the local Kubernetes API server to get the list of your configurations that are in your cluster.
-For more information about CRDs, see https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/[Extend the Kubernetes API with CustomResourceDefinitions].
+A Kubernetes object is defined (in whole or in part) in the `object-templates` array in the configuration policy, indicating to the configuration policy controller of the fields to compare with objects on the managed cluster. The configuration policy controller communicates with the local Kubernetes API server to get the list of your configurations that are in your cluster. For more information a ...
```
Why are CRDs relevant here?

dhaiducek replied:
I left those references in, but I agree they seem a bit too high-level for our docs. 

@dockerymick Do you have thoughts on the CRD links?

dockerymick replied:
@dhaiducek not really, just thought it would be helpful to add this reference in case a user was unfamiliar with Kubernetes. I am ok with removing this link reference

mprahl commented on governance/create_config_pol.adoc:
```diff
@@ -51,16 +41,21 @@ apiVersion: policy.open-cluster-management.io/v1
+    name: mustonlyhave-configuration
+  spec:
+    namespaceSelector:
+      include: ["default", "kube-*"]
```
The `kube-*` seems wrong since we usually never want a policy to target those namespaces.

mprahl commented on governance/policy_generator.adoc:
```diff
@@ -539,118 +539,119 @@ roleRef:
 
 Note that all the fields in the `policyDefaults` section except for `namespace` can be overridden per policy.
```
Optional: It'd be nice to confirm that all the options are documented here that are in the Policy Generator reference:
https://github.com/stolostron/policy-generator-plugin/blob/main/docs/policygenerator-reference.yaml

dhaiducek replied:
Updated--please take a peek! There's an experiment in the commit to have the root-level key span the row, but that can easily be updated if it's too unconventional.

mprahl replied:
@dhaiducek I like the change, but I'll let @dockerymick be the final judge on it.  :laughing:

dockerymick committed:
Update policy_generator.adoc

Adding a reference to the end of the page

mprahl commented:
Thanks so much @dhaiducek and @dockerymick! This change looks awesome.

mprahl commented on governance/policy_generator.adoc:
```diff
@@ -655,3 +655,5 @@ Note that all the fields in the `policyDefaults` section except for `namespace`
 | A Kustomize patch to apply to the manifest at the path. If there are multiple manifests, the patch requires the `apiVersion`, `kind`, `metadata.name`, and `metadata.namespace` (if applicable) fields to be set so Kustomize can identify the manifest that the patch applies to. If there is a single manifest, the `metadata.name` and `metadata.namespace` fields can be patched.
 |===
+
+Return to the xref:../governance/thirf_party_policy.adoc#integrate-third-party-policy-controllers[Integrate third-party policy controllers] documentation or refer to the xref:../governance/grc_intro.adoc#governance[Governance] documentation for more topics.
```
Seems like a typo `thirf_party_policy`

dockerymick committed:
Update psp_policy.adoc

Updates

dockerymick committed:
Update psp_policy.adoc

dockerymick committed:
Update role_policy.adoc

dhaiducek committed:
Overhaul generator table

- Add missing fields
- Add spanning rows for root-level keys (experiment)

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek committed:
Remove `kube-*` from `include` of ns selector

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek committed:
Update the introduction

- Clarify the name/namespace length restriction
- Remove reference to custom controllers

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dockerymick committed:
Update config_policy_ctrl.adoc

Removing CRD reference

dockerymick committed:
Update config_policy_ctrl.adoc

more updates

dockerymick committed:
Update create_config_pol.adoc

Updating reference link

dockerymick committed:
Update policy_generator.adoc

dockerymick committed:
Update rolebinding_policy.adoc

dockerymick committed:
Update rolebinding_policy.adoc

Updating conclusion

dockerymick committed:
Update scc_policy.adoc

updates

dockerymick committed:
Update rolebinding_policy.adoc

dhaiducek committed:
Update policy generator

- Adjust bold for parent fields
- Add `evaluationInterval` parameters

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek committed:
Remove CRD links

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek committed:
Add link to text for YAML table

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

oafischer commented on governance/cert_policy_ctrl.adoc:
```diff
@@ -1,7 +1,7 @@
 = Certificate policy controller
 
-Certificate policy controller can be used to detect certificates that are close to expiring, and detect time durations (hours) that are too long or contain DNS names that fail to match specified patterns.
+Certificate policy controller can be used to detect certificates that are close to expiring, time durations (hours) that are too long, or contain DNS names that fail to match specified patterns. The certificate policy is provided in the `policy-templates` field of the policy on the hub cluster and is propagated to the selected managed clusters by the governance framework. See the xref:../governan ...
```
...for more details on...

dhaiducek replied:
Updated!

dhaiducek committed:
for more about --> for more details on

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

oafischer commented on governance/config_policy_ctrl.adoc:
```diff
@@ -1,64 +1,21 @@
 = Kubernetes configuration policy controller
 
-Configuration policy controller can be used to configure any Kubernetes resource and apply security policies across your clusters.
+The configuration policy controller can be used to configure any Kubernetes resource and apply security policies across your clusters. The configuration policy is provided in the `policy-templates` field of the policy on the hub cluster, and is propagated to the selected managed clusters by the governance framework. See the xref:../governance/policy_example.adoc#policy-yaml-structure[policy YAML  ...
```
...for more details on...

oafischer commented on governance/config_policy_ctrl.adoc:
```diff
@@ -100,64 +59,69 @@ spec:
+`mustonlyhave`: Indicates that an object must exist with the exact fields and values as defined in the object.
 
-`musthave`: Indicates an object must exist with the same name as specified object-template. The other fields in the template are a subset of what exists in the object.
+`musthave`: Indicates an object must exist with the same fields as specified in the objectDefinition. The fields in the template are a subset of what exists in the object. In general, array values are appended. The exception for the array to be patched is when the item contains a `name` key with a value that matches an existing item. If the array should be replaced, use a fully defined `objectDef ...
```
This is more for future reference but we generally try to use active voice where possible (I know it's not always possible).

oafischer commented on governance/config_policy_ctrl.adoc:
```diff
@@ -100,64 +59,69 @@ spec:
+== Configure the configuration policy controller
 
-You can configure the concurrency of the configuration policy controller per managed cluster to change how many configuration policies it can evaluate at the same time. To change the default value of `2`, set the `policy-evaluation-concurrency` annotation with a non-zero integer within quotes. You can set the value on the `ManagedClusterAddOn` object called `config-policy-controller` in the manag ...
+You can configure the concurrency of the configuration policy controller per managed cluster to change how many configuration policies it can evaluate at the same time. To change the default value of `2`, set the `policy-evaluation-concurrency` annotation with a non-zero integer within quotes. You can set the value on the `ManagedClusterAddOn` object called `config-policy-controller` in the manag ...
```
-> You can configure the concurrency of the configuration policy controller for each managed cluster

(change for translation/accessibility)

dhaiducek replied:
Updated--thanks!

dhaiducek committed:
per --> for each

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

oafischer commented on governance/create_gatekeeper.adoc:
```diff
@@ -27,13 +27,13 @@ Gatekeeper operator policy is monitored by the {product-title-short} configurati
-*Note:* Consider that default values can be generated by the operator. See https://github.com/open-policy-agent/gatekeeper/blob/master/charts/gatekeeper/README.md[Gatekeeper Helm Chart] for an explanation of the optional parameters that can be used for the gatekeeper operator policy.  
+Select *Create policy*. As you complete the form, select *Gatekeeper Operator* from the _Specifications_ field. The parameter values for your policy are automatically populated and the policy is set to `inform` by default. Set your remediation action to `enforce` to install gatekeeper.
+
+*Note:* Default values are generated by the operator. See https://github.com/open-policy-agent/gatekeeper/blob/master/charts/gatekeeper/README.md[Gatekeeper Helm Chart] for an explanation of the optional parameters that can be used for the gatekeeper operator policy.  
```
For example, here active voice would be: The operator generates (the) default values.

dhaiducek replied:
I was just about to ask about an example of active voice--thanks!

oafischer commented on governance/grc_intro.adoc:
```diff
@@ -18,15 +18,15 @@ The governance architecture is composed of the following components:
 
 +
-** When a policy is propagated to a managed cluster, the replicated policy is named `namespaceName.policyName`. When you create a policy, make sure that the length of the `namespaceName.policyName` must not exceed 63 characters due to the Kubernetes limit for object names.  
+** When a policy is propagated to a managed cluster, it is first replicated to the cluster namespace on the hub cluster, is named `namespaceName.policyName`, and also labeled as such. When you create a policy, make sure that the length of the `namespaceName.policyName` does not exceed 63 characters due to the Kubernetes limit on the length of label values.
```
Can we say the following?: When a policy is propagated to a managed cluster, it is first replicated to the cluster namespace on the hub cluster and is named and labeled `namespaceName.policyName`.

(we try to avoid referring to things with this, that, as such etc., also not always possible but it might work here)

dhaiducek replied:
Yes--could we say "named and labeled using `namespaceName.policyName`"?

The reason is that `namespaceName.policyName` is used in the label as `policy.open-cluster-management.io/root-policy=namespaceName.policyName`, so it's not the label key.

oafischer replied:
Got it, I was thinking it might be something like that so feel free to keep it as is, if it's more accurate from a technical standpoint.

dockerymick committed:
Update policy_example.adoc

small updates

oafischer commented on governance/grc_intro.adoc:
```diff
@@ -18,15 +18,15 @@ The governance architecture is composed of the following components:
 
 +
-** When a policy is propagated to a managed cluster, the replicated policy is named `namespaceName.policyName`. When you create a policy, make sure that the length of the `namespaceName.policyName` must not exceed 63 characters due to the Kubernetes limit for object names.  
+** When a policy is propagated to a managed cluster, it is first replicated to the cluster namespace on the hub cluster, is named `namespaceName.policyName`, and also labeled as such. When you create a policy, make sure that the length of the `namespaceName.policyName` does not exceed 63 characters due to the Kubernetes limit on the length of label values.
```
...due to the Kubernetes length limit for label values.

oafischer commented on governance/memory_policy.adoc:
```diff
@@ -1,9 +1,7 @@
-Kubernetes configuration policy controller monitors the status of the memory usage policy.
-Use the memory usage policy to limit or restrict your memory and compute usage.
-For more information, see _Limit Ranges_ in the https://kubernetes.io/docs/concepts/policy/limit-range/[Kubernetes documentation].
+Kubernetes configuration policy controller monitors the status of the memory usage policy. Use the memory usage policy to limit or restrict your memory and compute usage. For more information, see _Limit Ranges_ in the https://kubernetes.io/docs/concepts/policy/limit-range/[Kubernetes documentation].
```
Does this need to start with "The Kubernetes..."? Saw it in one of the other files in this PR.

dhaiducek replied:
Oh, nice catch!

dhaiducek replied:
I'll go through and fix all of them--it's a common paragraph for the policy examples.

oafischer commented on governance/namespace_policy.adoc:
```diff
@@ -1,13 +1,12 @@
 
-Kubernetes configuration policy controller monitors the status of your namespace policy.
-Apply the namespace policy to define specific rules for your namespace.
+Kubernetes configuration policy controller monitors the status of your namespace policy. Apply the namespace policy to define specific rules for your namespace.
```
Same question as before about needing "The" in the beginning.

oafischer commented on governance/namespace_policy.adoc:
```diff
@@ -42,52 +51,39 @@ spec:
+| `spec.remediationAction`
 | Optional
-| Specifies the remediation of your policy. The parameter values are `enforce` and `inform`.
+| Specifies the remediation of your policy. The parameter values are `enforce` and `inform`. This value is optional since it will override any values provided in `spec.policy-templates`.
```
We also generally try to use present tense. So here I would change to: This value is optional because it overrides any values provided in `spec.policy-templates`.

oafischer commented on governance/policy_example.adoc:
```diff
@@ -82,56 +80,57 @@ spec:
+| Specifies the remediation of your policy. The parameter values are `enforce` and `inform`. If specified, the `spec.remediationAction` value that is defined overrides any `remediationAction` parameter defined in the child policies in the `policy-templates` section. For example, if the `spec.remediationAction` value is set to `enforce`, then the `remediationAction` in the `policy-templates` secti ...
 
-*Important*: Some policies might not support the enforce feature.
+*Important*: Some policy kinds might not support the enforce feature.
```
-> **Important:**

(make colon bold)

dhaiducek replied:
Does this go for all bold things like this? If so, there are a lot of updates to make (which I am totally primed to attack!)

oafischer replied:
This is only for special tags. See https://stolostron.github.io/rhacm-docs/wr_special_tag_guidance

dhaiducek replied:
Cool, thanks! I've created another commit to address all the instances: https://github.com/stolostron/rhacm-docs/pull/3870/commits/473a6abba1bb006a9edaefddbc44848598f9be94

oafischer replied:
Awesome, thanks! There are a lot of places in the doc where it is inconsistent so our goal has just been to fix it whenever we happen to see it while working on issues. This definitely helps.

dhaiducek replied:
If it helps, you can use VSCode to find/replace them all at once:
- Search for this RegEx (The `.*` option in Search): `((note|important|tip|optional|best practice)s?)\*:`
- Replace with the swapped asterisk and colon: `$1:*`

To explain, the stuff in parentheses gets captured (with an optional "s") and you can reference it using $1 (if there were more than one parentheses, the next group would be $2, and so on).

oafischer commented on governance/policy_generator.adoc:
```diff
@@ -539,118 +539,179 @@ roleRef:
+
+| `policyDefaults.evaluationInterval`
+| Optional
+| Use the parameters `compliant` and `noncompliant` to specify the frequency for a policy to be evaluated when in a particular compliance state. When managed clusters have low CPU resources, the evaluation interval can be increased to to reduce CPU usage on the Kubernetes API. These are in the format of durations (e.g. "1h25m3s"). These can also be set to "never" to avoid evaluating the policy af ...
```
Remove the extra "to"

oafischer commented on governance/policy_generator.adoc:
```diff
@@ -539,118 +539,179 @@ roleRef:
+
+| `policyDefaults.evaluationInterval`
+| Optional
+| Use the parameters `compliant` and `noncompliant` to specify the frequency for a policy to be evaluated when in a particular compliance state. When managed clusters have low CPU resources, the evaluation interval can be increased to to reduce CPU usage on the Kubernetes API. These are in the format of durations (e.g. "1h25m3s"). These can also be set to "never" to avoid evaluating the policy af ...
```
Do not use e.g.

IBM style says replace e.g. with "that is", though I'd prefer if you just write around that so you don't need to use either.

dhaiducek replied:
So would something like this work or is there a better way?
>These are in the format of durations like `"1h25m3s"`.

oafischer replied:
Hmm sometimes it's hard to write around. One option is to use a new sentence:

-> The interval is in the format of durations. For example, `1h25m3s` represents 1 hour, 25 minutes, and 3 seconds.

I guess in this case you could also use "for example" in the brackets instead of e.g.

oafischer commented on governance/policy_generator.adoc:
```diff
@@ -539,118 +539,179 @@ roleRef:
-| policyDefaults.placement
+| `policyDefaults.generatePlacementWhenInSet`
+| Optional
+| When a policy is part of a policy set, by default, the generator does not generate the placement for this policy since a placement is generated for the policy set. If a placement should still be generated, set it to `true` so that the policy is deployed with both policy placement and policy set placement. The default value is `false`.
```
Avoid using "should" by writing in a direct manner. I would change to something like: To generate a placement, set it to...

oafischer commented on governance/policy_generator.adoc:
```diff
@@ -539,118 +539,179 @@ roleRef:
+
+| `policySets[].placement`
+| Optional
+| The placement configuration for the policy set. This defaults to a placement configuration that matches all clusters. See `policyDefaults.placement` for placement documentation, however `policyDefaults.placement` settings do not get applied to policy sets.
```
...do not apply to policy sets.

(active voice)

oafischer reviewed (changes requested):
Thanks for the massive work on this! I've left a bunch of comments and suggestions.

dhaiducek committed:
Add "The" before config controller

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek committed:
Reflow replicated name explanation

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek committed:
Remove extra colon

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek committed:
Bold colon

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek committed:
Active voice

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek committed:
Boldify special tag colons

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dhaiducek committed:
Update iam_policy_ctrl.adoc

Signed-off-by: Dale Haiducek <dhaiduce@redhat.com>

dockerymick committed:
Update config_policy_ctrl.adoc

small updates

dockerymick committed:
Update config_policy_ctrl.adoc

dockerymick committed:
Update grc_intro.adoc

small update

dhaiducek commented:
I think this PR is ready for another review if everyone could take one (hopefully last) peek...

oafischer reviewed (approved):
Looks good from a writing perspective. Thank you for also bolding the colons for special tags!

dhaiducek commented:
/unhold

dockerymick committed:
Update cert_policy_ctrl.adoc

Fixing link

dockerymick committed:
Update config_policy_ctrl.adoc

Fixing link

dockerymick committed:
Update etcd_encryption_policy.adoc

fixing link

dockerymick committed:
Update iam_policy_ctrl.adoc

fixing link

dockerymick committed:
Update memory_policy.adoc

dockerymick committed:
Update cert_policy_ctrl.adoc

consistency

dockerymick committed:
Update config_policy_ctrl.adoc

dockerymick committed:
Update namespace_policy.adoc

consistency

dockerymick committed:
Update pod_policy.adoc

dockerymick committed:
Update policy_generator.adoc

dockerymick committed:
Update psp_policy.adoc

updating link and page name

dockerymick committed:
Update rolebinding_policy.adoc

dockerymick committed:
Update scc_policy.adoc

dockerymick reviewed (approved):
Wheew, all files viewed and links checked in this PR

dockerymick committed:
Merge branch '2.6_stage' into refresh-docs

dockerymick commented:
@dhaiducek confirmed that he is ok with me merging this PR through Slack.
